ostd/src/task/task.rs

```diff
@@ -42,65 +45,56 @@ pub trait TaskContextApi {
 }
 
 pub struct KernelStack {
-    segment: Segment,
+    kva: Kva,
+    mapped: Range<Vaddr>,
     has_guard_page: bool,
 }
 
 impl KernelStack {
     pub fn new() -> Result<Self> {
+        let mut kva_alloc = Kva::alloc(KERNEL_STACK_SIZE);
+        let mapped_start = kva_alloc.start();
+        let mapped_end = kva_alloc.end();
+        // let pages = allocator::alloc::<FrameMeta>(KERNEL_STACK_SIZE).unwrap();
+        // let pages = allocator::alloc_contiguous::<KernelStackMeta>((KERNEL_STACK_SIZE)).unwrap();
+        let pages = allocator::alloc::<KernelStackMeta>(KERNEL_STACK_SIZE).unwrap();
+        unsafe { kva_alloc.map_pages(mapped_start..mapped_end, pages) }
         Ok(Self {
-            segment: FrameAllocOptions::new(KERNEL_STACK_SIZE / PAGE_SIZE).alloc_contiguous()?,
+            kva: kva_alloc,
+            mapped: mapped_start..mapped_end,
             has_guard_page: false,
         })
     }
 
     /// Generates a kernel stack with a guard page.
     /// An additional page is allocated and be regarded as a guard page, which should not be accessed.  
     pub fn new_with_guard_page() -> Result<Self> {
-        let stack_segment =
-            FrameAllocOptions::new(KERNEL_STACK_SIZE / PAGE_SIZE + 1).alloc_contiguous()?;
-        // FIXME: modifying the the linear mapping is bad.
-        let page_table = KERNEL_PAGE_TABLE.get().unwrap();
-        let guard_page_vaddr = {
-            let guard_page_paddr = stack_segment.start_paddr();
-            crate::mm::paddr_to_vaddr(guard_page_paddr)
-        };
-        // SAFETY: the segment allocated is not used by others so we can protect it.
+        let mut kva_alloc = Kva::alloc(KERNEL_STACK_SIZE + 4 * PAGE_SIZE);
+        let mapped_start = kva_alloc.start() + 2 * PAGE_SIZE;
+        let mapped_end = mapped_start + KERNEL_STACK_SIZE;
+        // let frames = FrameAllocOptions::new(KERNEL_STACK_SIZE).uninit(true).alloc()?;
+        let pages = allocator::alloc::<KernelStackMeta>(KERNEL_STACK_SIZE).unwrap();
+        // let pages = allocator::alloc_contiguous::<KernelStackMeta>((KERNEL_STACK_SIZE)).unwrap();
+        // let pages = allocator::alloc::<FrameMeta>(KERNEL_STACK_SIZE).unwrap();
         unsafe {
-            let vaddr_range = guard_page_vaddr..guard_page_vaddr + PAGE_SIZE;
-            page_table
-                .protect(&vaddr_range, |p| p.flags -= PageFlags::RW)
-                .unwrap();
-            tlb_flush_addr_range(&vaddr_range);
+            kva_alloc.map_pages(mapped_start..mapped_end, pages);
         }
         Ok(Self {
-            segment: stack_segment,
+            kva: kva_alloc,
+            mapped: mapped_start..mapped_end,
             has_guard_page: true,
         })
     }
 
-    pub fn end_paddr(&self) -> Paddr {
-        self.segment.end_paddr()
+    pub fn end_vaddr(&self) -> Vaddr {
+        self.mapped.end
     }
 }
 
 impl Drop for KernelStack {
```

Comment:
Unmapping should be implemented in the dropping method of `KvaInner`, so there would be no need implementing `Drop` for `KernelStack` here.